### ISSUE TITLE
Parsing JWTs with a private claim called "payload" fails

### ISSUE THREAD

**Issue by valerio** [lestrrat-go/jwx#430]
In my code I sometimes have to parse JWTs that contain a private claim called `payload`, for example:
```json
{
  "sub": "some-id",
  "payload": {
    "firstName": "John",
    "lastName": "Smith"
  }
}
```

This kind of token used to be parsed without issues on previous versions of the library, but after upgrading to version 1.2.2 (or above), calling `jwt.Parse()` on a token like this results in an error (formatted with extra newlines):
```
failed to verify jws signature: 
failed to unmarshal JSON message: 
failed to unmarshal into temporary structure: 
json: cannot unmarshal object into Go struct field messageProxy.payload of type string
```

Output of `go version`:
```
go version go1.16.4 darwin/amd64
```

**To Reproduce / Expected behavior**

Here is a simple test that reproduces the issue (still present as of version 1.2.5):
```go
func TestTokenWithPayloadClaim(t *testing.T) {
	t1 := jwt.New()
	err := t1.Set("payload", map[string]interface{}{
		"name": "someone",
	})
	if err != nil {
		t.Fatalf("JWT set error: %s", err.Error())
	}

	bytes, err := jwt.Sign(t1, jwa.HS256, []byte("secret"))
	if err != nil {
		t.Fatalf("JWT sign error: %s", err.Error())
	}

	_, err = jwt.Parse(bytes, jwt.WithVerify(jwa.HS256, []byte("secret")))
	if err != nil {
		t.Fatalf("JWT parse error: %s", err.Error())
	}
}
```

I would expect `jwt.Parse` to return no error, which is what happens on version 1.2.1 or below.

**Additional context**

This issue seems to have started happening from [version 1.2.2](https://github.com/lestrrat-go/jwx/releases/tag/v1.2.2).
I noticed that this release added an extra step where the library tries to guess the format of the token by unmarshaling it into a temporary JSON object. The temporary struct [uses a `payload` field](https://github.com/lestrrat-go/jwx/compare/v1.2.1...v1.2.2#diff-4e36647f41c8fcc4fe2791fec1c379ac314053900486e3eb80e116b1c20ea731R20) when unmarshaling, which might be what's causing the issue.

I'm not quite sure yet where and how this parsing is failing (the error is actually mentioning [the `messageProxy` struct](https://github.com/lestrrat-go/jwx/blob/main/jws/message.go#L159), which also has a `payload` field).

I've checked the [JWS RFC](https://datatracker.ietf.org/doc/html/rfc7515#section-7.2.1) and it does mention the `payload` field being a required member of its JSON serialization format. I'm not sure if this applies when parsing a JWT's payload though, but obviously, if this is a case of the library now sticking closer to the RFC, I think we can close this issue.

If it is an actual issue, I'm happy to contribute a fix, although I'm not sure yet what that would be, I probably need to understand the parsing process a bit more.

**Comment by lestrrat** [comment]
Ah, I see. I started allowing JWE(JWS) messages, and I needed to guess what the format of whatever blob we got.
For JWS, I did it by detecting the "payload" field, and hence it thought that we had a `JWS-compact(JWS-JSON)` message.

Your payload was unintentionally detected as a JWS because you lack the `"aud"`  (which tells us it's a JWT), and you had `"payload"` (which tells us it's a JWS). This was probably avoidable if you had a `"aud"` claim in your JWT.

In light of this I've changed the detection heuristics to check for the presence of _both_ `"signatures"` and `"payload"` to determine it's a JWS message. However, this can happen again if you don't use the `"aud"` claim and you contain some other field name which clashes with other "hint" fields that we use.

Please test #431 and see if this fixes your problem

**Comment by lestrrat** [comment]
Oh, I forgot to mention, `"aud"` is mandatory for JWTs! :)

**Comment by valerio** [comment]
@lestrrat can confirm https://github.com/lestrrat-go/jwx/pull/431 fixes the issue for me, thank you so much! 🥇 

And very good point on the `aud` claim, I'll make sure to add it in to avoid any other problem.

**Comment by lestrrat** [comment]
Thanks for checking!

**Comment by lestrrat** [comment]
fixed via #431
